File: Python/Pack.py
from Date import Date

class Pack:
    DEFAULT_RECEIVER = "Default Company"
    DEFAULT_ZONE = "A1"
    DEFAULT_DATE = Date()
    DEFAULT_WEIGHT = 1
    DEFAULT_VOLUME = 1
    count = 1

    def __init__(self, receiver=None, zone=None,
                 date=None, weight=None,
                 volume=None):
        self.packageID = Pack.count
        self.set_receiver(receiver)
        self.set_zone(zone)
        self.set_delivery_date(date)
        self.set_weight(weight)
        self.set_volume(volume)
        Pack.packageID = Pack.count
        Pack.count += 1


    def get_receiver(self):
        return self.receiver

    def get_zone(self):
        return self.zone

    def get_delivery_date(self):
        return self.date

    def get_weight(self):
        return self.weight

    def get_volume(self):
        return self.volume

    def get_packageID(self):
        return self.packageID

    def get_time(self):
        return 17

    def set_receiver(self, name):
        if name is not None:
            self.receiver = name
        else:
            self.receiver = Pack.DEFAULT_RECEIVER

    def set_zone(self, z):
        if z is not None and "A" <= z[0] <= "Z" and "1" <= z[1] <= "9":
            self.zone = z
        else:
            self.zone = Pack.DEFAULT_ZONE

    def set_delivery_date(self, d):
        if d is not None and isinstance(d, Date):
            self.date = d
        else:
            self.date = Date(Date.DEFAULT_MONTH, Date.DEFAULT_DAY, Date.DEFAULT_YEAR)

    def set_weight(self, w):
        if w is not None:
            if w > 0:
                self.weight = w
            else:
                self.weight = Pack.DEFAULT_WEIGHT
        else:
            self.weight = Pack.DEFAULT_WEIGHT

    def set_volume(self, v):
        if v is not None:
            if v > 0:
                self.volume = v
            else:
                self.volume = Pack.DEFAULT_VOLUME
        else:
            self.volume = Pack.DEFAULT_VOLUME

    def __str__(self):
        return f"ID:{self.get_packageID()} Company:{self.receiver} Zone:{self.zone} Date:{self.date} Weight:{self.weight} Volume:{self.volume}"

    def compare_to(self, pack):
        if self.get_zone() < pack.get_zone():
            return 1
        elif self.get_zone() > pack.get_zone():
            return -1
        else:
            if self.get_volume() > pack.get_volume():
                return 1
            elif self.get_volume() < pack.get_volume():
                return -1
            else:
                return 0

File: CPP/Pack.cpp
#include "headers/Pack.h"

int Pack::count = 0;

// default constructor;
Pack::Pack() {
    receiver = DEFAULT_RECEIVER;
    zone = DEFAULT_ZONE;
    deliveryDate = DEFAULT_DATE;
    weight = DEFAULT_WEIGHT;
    volume = DEFAULT_VOLUME;
    packageID = count;
    count++;
}

// non-default constructor
Pack::Pack(const std::string& r, const std::string& z, const Date& d, int w, int v) {
    receiver = r;
    zone = z;
    deliveryDate = d;
    weight = w;
    volume = v;
    packageID = count;
    count++;
}

// accessor methods
std::string Pack::getReceiver() const {return receiver;}
std::string Pack::getZone() const {return zone;}
Date Pack::getDeliveryDate() const {return deliveryDate;}
int Pack::getWeight() const {return weight;}
int Pack::getVolume() const {return volume;}
int Pack::getPackageID() const {return packageID;}
int Pack::getTime() {return 17;}

// mutator methods
void Pack::setReceiver(const std::string& n) {
    if (!n.empty() ) {
        receiver = n;
    }
    else receiver = DEFAULT_RECEIVER;
}

void Pack::setZone(const std::string& z) {
    if (!z.empty() && z.at(0) >= 'A' && z.at(0) <= 'Z' &&
        z.at(1) >= '1' && z.at(1) <= '9' ) {
        this->zone = z;
    }
    else this->zone = DEFAULT_ZONE;
}

void Pack::setDeliveryDate(const Date& d) {
    if (d.month > 0 && d.day > 0 && d.year > 0) {
        deliveryDate = d;
    }
    else deliveryDate = DEFAULT_DATE;
}

void Pack::setWeight(int w) {
    if (w > 0) {
        weight = w;
    }
    else weight = DEFAULT_WEIGHT;
}

void Pack::setVolume(int v) {
    if (v > 0) {
        volume = v;
    }
    else volume = DEFAULT_VOLUME;
}

// toString
std::string Pack::toString() {
    return "ID: " + std::to_string(getPackageID()) + " Company: " + getReceiver() + " Zone: " + getZone() +
        " Date: " + getDeliveryDate().toString() + " Weight: " + std::to_string(getWeight()) + " Volume: " + std::to_string(getVolume());
}

// comparison
int Pack::compareTo(const Pack& that) const {
    if (this->getZone()[0] == that.getZone()[0]) {
        if (this->getZone()[1] == that.getZone()[1]) {
            if (this->getVolume() > that.getVolume()) {
                return 1;
            }
            else if (this->getVolume() < that.getVolume()) {
                return -1;
            }
            else return 0;
        }
        else if (this->getZone()[1] < that.getZone()[1]) {
            return 1;
        }
        else return -1;
    }
    else if (this->getZone()[0] > that.getZone()[0]) {
        return -1;
    }
    else if (this->getZone()[0] < that.getZone()[0]) {
        return 1;
    }
    else return 0;
}

// overloading operators
bool Pack::operator==(const Pack& other) const {
    return (packageID == other.packageID);
}

std::ostream& operator<<(std::ostream& os, const Pack& p) {
    os << ": Pack(ID: " << p.getPackageID() << ", Company: " << p.getReceiver()
        << ", Zone: "<< p.getZone() << ", Date: " << p.getDeliveryDate().toString()
        << ", Weight: " << p.getWeight() << ", Volume: " << p.getVolume() << ")";
    return os;
}

